src/mesh/boundary_standard.cxx

```diff
@@ -525,454 +218,238 @@ void BoundaryDirichlet::apply(Field3D &f,BoutReal t) {
         // Set any other guard cells using the values on the cells
         int xi = bndry->x + i*bndry->bx;
         int yi = bndry->y + i*bndry->by;
-        xnorm = mesh->GlobalX(xi);
-        ynorm = mesh->GlobalY(yi);
-        for(int zk=0;zk<mesh->LocalNz;zk++) {
+        xnorm = localmesh->GlobalX(xi);
+        ynorm = localmesh->GlobalY(yi);
+        for(int zk=0; zk<f.getNz() ; zk++) {
+          BoutReal znorm = localmesh->GlobalZ(zk);
           if(fg) {
-            val = fg->generate(xnorm,TWOPI*ynorm,TWOPI*zk/(mesh->LocalNz), t);
+            val = fg->generate(xnorm,TWOPI*ynorm,TWOPI*znorm, t);
           }
           f(xi, yi, zk) = val;
         }
       }
     }
-  }
-}
+  } else if( loc == CELL_XLOW ) {
+    // Field is shifted in X
+    if(bndry->bx > 0) {
+      // Outer x boundary
+      for(; !bndry->isDone(); bndry->next1d()) {
+        BoutReal xnorm = 0.5*(   localmesh->GlobalX(bndry->x)
+            + localmesh->GlobalX(bndry->x - bndry->bx) );
+        BoutReal ynorm = localmesh->GlobalY(bndry->y);
+
+        for(int zk=0; zk<f.getNz(); zk++) {
+          BoutReal znorm = localmesh->GlobalZ(zk);
+          if(fg){
+            val = fg->generate(xnorm,TWOPI*ynorm,TWOPI*znorm, t);
+          }
+          f(bndry->x,bndry->y, zk) = val;
 
+          // Need to set second guard cell, as may be used for interpolation or upwinding derivatives
+          for(int i=1;i<bndry->width;i++) {
+            int xi = bndry->x + i*bndry->bx;
+            int yi = bndry->y ;
 
-void BoundaryDirichlet::apply_ddt(Field2D &f) {
-  Field2D *dt = f.timeDeriv();
-  for(bndry->first(); !bndry->isDone(); bndry->next())
-    (*dt)(bndry->x,bndry->y) = 0.; // Set time derivative to zero
-}
+            f(xi, yi, zk) = 2*f(xi - bndry->bx, yi , zk) - f(xi- 2*bndry->bx, yi , zk);
+          }
+        }
+      }
+    }
+    if (bndry->bx < 0){
```

Comment:
While we're fixing things here, this should be `else if` -- `bndry.bx` can't be both `< 0` and `> 0`